olp-cpp-sdk-dataservice-read/src/StreamLayerClientImpl.cpp

```diff
@@ -280,6 +280,101 @@ client::CancellableFuture<DataResponse> StreamLayerClientImpl::GetData(
                                                       std::move(promise));
 }
 
+client::CancellationToken StreamLayerClientImpl::Poll(
+    PollResponseCallback callback) {
+  auto poll_task = [=](client::CancellationContext context) -> PollResponse {
+    std::string subscription_id;
+    std::string subscription_mode;
+    std::string x_correlation_id;
+    std::shared_ptr<client::OlpClient> client;
+
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      if (!client_context_) {
+        OLP_SDK_LOG_WARNING_F(kLogTag,
+                              "Poll: unsuccessful, not subscribed");
+
+        return client::ApiError(client::ErrorCode::PreconditionFailed,
+                                "Not subscribed", false);
+      }
+
+      subscription_id = client_context_->subscription_id;
+      subscription_mode = client_context_->subscription_mode;
+      x_correlation_id = client_context_->x_correlation_id;
+      client = client_context_->client;
+    }
+
+    OLP_SDK_LOG_INFO_F(kLogTag,
+                       "Poll: started, subscription_id=%s, "
+                       "subscription_mode=%s, x_correlation_id=%s",
+                       subscription_id.c_str(), subscription_mode.c_str(),
+                       x_correlation_id.c_str());
+
+    auto data =
+        StreamApi::ConsumeData(*client, layer_id_, subscription_id,
+                               subscription_mode, context, x_correlation_id);
+
+    if (!data.IsSuccessful()) {
+      OLP_SDK_LOG_WARNING_F(kLogTag, "Poll: consume data unsuccessful, error=%s",
+                            data.GetError().GetMessage().c_str());
+      return data.GetError();
+    }
+
+    auto res = data.MoveResult();
+
+    const auto& messages = res.GetMessages();
+
+    if (!messages.empty()) {
+      // Get offsets for all partitions presented in messages.
+      std::vector<model::StreamOffset> offsets;
+      for (const auto& message : messages) {
+        const auto current_offset = message.GetOffset();
+        auto iter = std::find_if(
```

Comment:
Hard to comment without numbers at hand but asymptotically this algorithm is not performant enough.
Is the expected number of messages and offsets small enough, so we're okay with having an O(n^2) algorithm?

Reply:
Changed